### PR TITLE
Add dfs flag for `typeclasses eauto`

### LINKED ISSUE #13859: Feature request: dfs flag for the `typeclasses eauto` tactic
#### Description of the problem

AFAICT there is no way to locally undo a global `Typeclasses eauto := (bfs)`: while one can write `typeclasses eauto bfs with xyz` to override the global default `dfs`, there no corresponding `typeclasses eauto dfs with xyz` to override a global `bfs` flag.

#### Coq Version

8.13

### PR DESCRIPTION
**Pull request by Alizter**
Fixes / closes #13859

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  - [x] Updated **documented syntax** by running `make -f Makefile.make doc_gram_rsts`.

### CODE REVIEW

**Comment by Alizter** [comment]
Should I update the documentation to talk about the new syntax?

File: doc/changelog/04-tactics/14693-alizter+typeclasses-eauto-dfs-flag.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Added:**
+  `typeclasses eauto dfs` syntax which locally enables depth-first search analogously to the `typeclasses eauto bfs` syntax.
```
Comment (jfehrle) [inline doc/changelog/04-tactics/14693-alizter+typeclasses-eauto-dfs-flag.rst:2]:
```suggestion
  `dfs` option in :tacn:`typeclasses eauto` to use depth-first search
```

It's convenient for readers to have the hyperlink to the tactic.

File: doc/tools/docgram/orderedGrammar [hunk 2]
```diff
@@ -1742,6 +1742,8 @@ simple_tactic: [
 | "autounfold" OPT hintbases OPT simple_occurrences
 | "autounfold_one" OPT hintbases OPT ( "in" ident )
 | "unify" one_term one_term OPT ( "with" ident )
+| "typeclasses" "eauto" "dfs" OPT nat_or_var "with" LIST1 ident
+| "typeclasses" "eauto" "dfs" OPT nat_or_var
```
Comment (jfehrle) [inline doc/tools/docgram/orderedGrammar:1746]: For the doc, these added lines should merged into the existing syntax for the tactic by editing `common.edit_mlg` like this (not tested):

Replace

```
| "typeclasses" "eauto" OPT "bfs" OPT nat_or_var OPT ( "with" LIST1 preident )
```

with

```
| "typeclasses" "eauto" OPT ( "bfs" | "dfs ) OPT nat_or_var OPT ( "with" LIST1 preident )
| DELETE "typeclasses" "eauto" "dfs" OPT nat_or_var "with" LIST1 preident
| DELETE "typeclasses" "eauto" "dfs" OPT nat_or_var
```

then run `make doc_gram_rsts`, which will update the syntax in the rst files.

Also if you would describe both bfs and dfs in the tactic description just above these lines (see the `Typeclasses eauto` command description of these) and with the same format (e.g. ```:n:`{| bfs | dfs }```):

```
   :n:`@nat_or_var`
     Specifies the maximum depth of the search.
```

**Review by jfehrle — CHANGES_REQUESTED** [review]
Please update as described.  The wording is a suggestion.

File: doc/sphinx/addendum/type-classes.rst [hunk 3]
```diff
@@ -452,6 +452,11 @@ Summary of the commands
    + Use the :cmd:`Typeclasses eauto` command to customize the behavior of
      this tactic.
 
+   :n:`{| bfs | dfs }
```
Comment (jfehrle) [inline doc/sphinx/addendum/type-classes.rst:455]:
```suggestion
   :n:`{| bfs | dfs }`
```

File: doc/sphinx/addendum/type-classes.rst [hunk 4]
```diff
@@ -452,6 +452,11 @@ Summary of the commands
...
+     Sets the search strategy to depth-first
+     search (the default) or breadth-first search. The search strategy
+     can also be set with :flag:`Typeclasses Iterative Deepening`.
```
Comment (jfehrle) [inline doc/sphinx/addendum/type-classes.rst:458]:
```suggestion
     Specifies whether to use breadth-first search or depth-first search.  The default is depth-first search,
     which can be changed with the :flag:`Typeclasses Iterative Deepening` flag.
```

"Sets" suggests that a change the persists after the tactic.
Isn't the default determined by the `Typeclasses Iterative Deepening`?
If that's correct.

And, if so, perhaps you would also tweak the wording for `Typeclasses eauto` (change the label to ```:n:`{| bfs | dfs }` ``` and swap the underlined wording so the order of the descriptions matches the order in the label

image

**Review by jfehrle — COMMENTED** [review]
Better, a few more nits.

File: doc/sphinx/addendum/type-classes.rst (hunk 4 above)
Comment (Alizter) [reply doc/sphinx/addendum/type-classes.rst:458]: Yes, I copied this from further down, I'll change both.

File: doc/sphinx/addendum/type-classes.rst [hunk 5]
```diff
@@ -452,6 +452,12 @@ Summary of the commands
    + Use the :cmd:`Typeclasses eauto` command to customize the behavior of
      this tactic.
 
+   :n:`{| bfs | dfs }`
+     Specifies whether to use depth-first search or breadth-first search.
```
Comment (jfehrle) [inline doc/sphinx/addendum/type-classes.rst:456]:
```suggestion
     Specifies whether to use breadth-first search or depth-first search.
```

File: doc/sphinx/addendum/type-classes.rst [hunk 6]
```diff
@@ -632,10 +638,10 @@ Typeclasses eauto
      successfully applied tactics is printed. Debug mode can also
      be set with :flag:`Typeclasses Debug`.
 
-   ``dfs``, ``bfs``
-     Sets the search strategy to depth-first
-     search (the default) or breadth-first search. The search strategy
-     can also be set with :flag:`Typeclasses Iterative Deepening`.
+   :n:`{| bfs | dfs }
+     Specifies whether to use depth-first search or breadth-first search.
```
Comment (jfehrle) [inline doc/sphinx/addendum/type-classes.rst:642]:
```suggestion
   :n:`{| bfs | dfs }`
     Specifies whether to use breadth-first search or depth-first search.
```

Don't forget the closing ``` ` ```.

**Review by jfehrle — COMMENTED** [review]
2 more small tweaks.

File: doc/sphinx/addendum/type-classes.rst (hunk 5 above)
Comment (Alizter) [reply doc/sphinx/addendum/type-classes.rst:456]: Ah, I swapped the order of the correct one before whoops.

**Review by jfehrle — APPROVED** [review]
The doc looks good

File: plugins/ltac/g_class.mlg [hunk 7]
```diff
@@ -86,12 +86,16 @@ END
 
 (** Compatibility: typeclasses eauto has 8.5 and 8.6 modes *)
```
Comment (Zimmi48) [inline plugins/ltac/g_class.mlg:87]: FWIW, this comment is outdated and could be removed (I don't know why it was located there in the first place).

**Review by Zimmi48 — COMMENTED** [review]
It's great to include a test case, but ideally, this test case would check that the two modes actually do something different (right now, you can only manually inspect the debug output). The idea would be to choose goals which fail when run with the wrong mode (`dfs` or `bfs`) and succeed with the other. Then you could insert `Fail 1: typeclasses eauto ...` to assert this.

Otherwise, LGTM :smile:

**Comment by Alizter** [comment]
@Zimmi48 They shouldn't succeed on the wrong modes in the test case. The classes and values were picked carefully so that they fail when the wrong mode is given.

File: plugins/ltac/g_class.mlg (hunk 7 above)
Comment (Alizter) [reply plugins/ltac/g_class.mlg:87]: OK, I will remove.

**Comment by Zimmi48** [comment]
That's what I expected when looking at the test case, so I wanted to suggest inserting the appropriate `Fail` statements. But when testing locally (with Coq 8.13.2), I was surprised to discover that the other mode also succeeded in these cases.

**Comment by Alizter** [comment]
Oh, that's not supposed to happen. Let me take a look.

**Comment by Alizter** [comment]
@Zimmi48 I'm confused, how are you testing this on 8.13? Could you give me something that doesn't work? I am unable to break it.

**Comment by Alizter** [comment]
In the test, `typeclasses eauto` will always succeed, its the following `reflexivity` that checks if it is correct. I've added some comments to the test to make it a bit clearer.

**Comment by Alizter** [comment]
In the test, `typeclasses eauto` will always succeed, its the following `reflexivity` that checks if it is correct. I've added some comments to the test to make it a bit clearer.

**Comment by Alizter** [comment]
In the test, `typeclasses eauto` will always succeed, its the following `reflexivity` that checks if it is correct. I've added some comments to the test to make it a bit clearer.

**Comment by Zimmi48** [comment]
>how are you testing this on 8.13?

I was testing without the new syntax. But indeed, I hadn't really read the test precisely and I had missed that the `reflexivity` tactic was the one that should fail. Your added explanations are great!

**Review by Zimmi48 — APPROVED** [review]
I'll merge this PR tomorrow.

**Comment by Zimmi48** [comment]
@coqbot merge now

**Pull request merged**